spring-integration-core/src/main/java/org/springframework/integration/handler/AbstractMessageProducingHandler.java

```diff
@@ -150,8 +167,54 @@ else if (reply instanceof AbstractIntegrationMessageBuilder) {
 			}
 		}
 
-		Message<?> replyMessage = createOutputMessage(reply, requestHeaders);
-		sendOutput(replyMessage, replyChannel);
+		if (this.asyncReplySupported && reply instanceof ListenableFuture<?>) {
+			ListenableFuture<?> future = (ListenableFuture<?>) reply;
+			final Object theReplyChannel = replyChannel;
+			future.addCallback(new ListenableFutureCallback<Object>() {
+
+				@Override
+				public void onSuccess(Object result) {
+					try {
+						sendOutput(createOutputMessage(result, requestHeaders), theReplyChannel, false);
+					}
+					catch (Exception e) {
+						Exception exceptionToLog = e;
+						if (!(e instanceof MessagingException)) {
+							exceptionToLog = new MessageHandlingException(requestMessage, e);
+						}
+						logger.error("Failed to send async reply: " + result.toString(), exceptionToLog);
```

Comment:
Well, actually for the upstream flow it is just a black-box how the handler behaves. And it really should look similar if we have there just one-way handler, or reply-producing. Just because the `sendOutput()` can be `sync` and some its handler behind may throw exception, the upstream should wait for the exception in its `try...catch`, if `sync` flow, or `ErrorMessage` from its `errorChannel` header, this `ListenableFuture`.

Just for more info. I see that you agree with me already. And I agree with you about the proper error message for this case.

And looks like we really don't have choice, unless `logger.error()` if there is no `errorChannel` header.
Although maybe eventually with reactive support we may consider an `error-channel` option from the config for these `ListenableFuture` cases, like we have in case of `<poller>`...